src/index.ts

```diff
@@ -85,12 +87,16 @@ const zodToTsNode = (
       const properties = Object.entries(zod._def.shape())
 
       const members: ts.TypeElement[] = properties.map(([key, value]) => {
-        const type = zodToTsNode(value as ZodTypeAny, ...otherArgs)
+        const zodAny = value as ZodTypeAny
+        const type = zodToTsNode(zodAny, ...otherArgs, { parentIsObject: true })
 
+        const flagsRequireOptional = options.treatOptionalsAs === 'both' || options.treatOptionalsAs === 'optional'
         return f.createPropertySignature(
           undefined,
           f.createIdentifier(key),
-          undefined,
+          zodAny._def.typeName === 'ZodOptional' && flagsRequireOptional
+            ? f.createToken(ts.SyntaxKind.QuestionToken)
+            : undefined,
```

Comment:
edge case with `zodAny._def.typeName === 'ZodOptional'`:
```ts
const schema = z.object({
  a: z.number().optional().transform((arg) => arg),
})

const { node } = zodToTs(schema, 'hello', {
  treatOptionalsAs: 'optional',
})

console.log(printNode(node))
```

result:
```ts
{
    a: number | undefined;
}
```

Reply:
Why won't we just remove `| undefined` afterwards if needed? 
See latest commit. I've updated tested with the cases provided above. They are still failing, but that's right direction, I think. Now we just need to define when to remove `| undefined` properly.